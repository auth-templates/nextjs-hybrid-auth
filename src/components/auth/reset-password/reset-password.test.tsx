import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ResetPassword from './reset-password';

describe('ResetPassword', () => {
    test('if all elements are present', () => {
        render(<ResetPassword onReset={jest.fn()}/>);
        expect(screen.getByLabelText('Password')).toBeInTheDocument();
        expect(screen.getByLabelText('Confirm password')).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Enter captcha")).toBeInTheDocument();
        expect(screen.getByRole('button', { name: "Reset password" })).toBeInTheDocument();
    })

    test('if the right values are received by onReset callback', () => {
        const mockOnReset = jest.fn();
        render(
            <ResetPassword 
                onReset={mockOnReset} 
            />
        );

        const passwordInput = screen.getByLabelText('Password'); 
        const confirmPasswordInput = screen.getByLabelText('Confirm password');
        const capchaInput = screen.getByPlaceholderText("Enter captcha");
        const resetButton = screen.getByRole('button', { name: "Reset password" });
        userEvent.type(passwordInput, '@Abcdef8');
        userEvent.type(confirmPasswordInput, '@Abcdef8');
        userEvent.type(capchaInput, 'w93bx');
        userEvent.click(resetButton);

        expect(mockOnReset).toBeCalledWith({ 
            password: "@Abcdef8",
            confirmPassword: "@Abcdef8",
            captcha: "w93bx",
        });
    })

    test('if message box is displayed when an error is received from the server', () => {
        render(
            <ResetPassword 
                onReset={jest.fn()} 
                status={{theme:'error', lines:['Failed to reset password. Please try again!']}}
            />
        );

        expect(screen.getByText(/Failed to reset password. Please try again!/)).toBeInTheDocument();
    })

    test('if password updated component is displayed when "Password updated" message is received', () => {
        render(
            <MemoryRouter>
                <ResetPassword 
                    onReset={jest.fn()} 
                    status={{theme:'info', lines:['Password updated']}}
                />
            </MemoryRouter>
        );

        expect(screen.getByText(/Your password has been reset successfully. Please go to/)).toBeInTheDocument();
    })

    test('if reset password email expired component is displayed when "Reset token is not valid message is received', () => {
        render(
            <MemoryRouter>
                <ResetPassword 
                    onReset={jest.fn()} 
                    status={{theme:'info', lines:['Reset token is not valid']}}
                />
            </MemoryRouter>
        );

        expect(screen.getByText(/Your reset password email has expired. Please go to/)).toBeInTheDocument();
    })

    test('if password inputs have theme danger and a message is displayed when a password is invalid', () => {
        render(
            <ResetPassword 
                onReset={jest.fn()}
            />
        );

        const passwordInput = screen.getByLabelText('Password'); 
        const confirmPasswordInput = screen.getByLabelText('Confirm password');
        const capchaInput = screen.getByPlaceholderText("Enter captcha");
        const resetButton = screen.getByRole('button', { name: "Reset password" });
        userEvent.type(passwordInput, '@Abcdef');
        userEvent.type(confirmPasswordInput, '@Abcdef');
        userEvent.type(capchaInput, 'w93bx');
        userEvent.click(resetButton);

        expect(passwordInput).toHaveClass('danger');
        expect(confirmPasswordInput).toHaveClass('danger');
        expect(screen.getByText(/You password should be at least 8 characters long and it should contain at least one lowercase character, one uppercase character, one symbol and one number./)).toBeInTheDocument();
    })

    test('if password inputs have theme danger and a message is displayed when passwords do no match', () => {
        render(
            <ResetPassword 
                onReset={jest.fn()}
            />
        );

        const passwordInput = screen.getByLabelText('Password'); 
        const confirmPasswordInput = screen.getByLabelText('Confirm password');
        const capchaInput = screen.getByPlaceholderText("Enter captcha");
        const resetButton = screen.getByRole('button', { name: "Reset password" });
        userEvent.type(passwordInput, '@Abcdef9');
        userEvent.type(confirmPasswordInput, '@Abcdef');
        userEvent.type(capchaInput, 'w93bx');
        userEvent.click(resetButton);

        expect(passwordInput).toHaveClass('danger');
        expect(confirmPasswordInput).toHaveClass('danger');
        expect(screen.getByText(/Passwords do not match/)).toBeInTheDocument();
    })

    it('renders correctly', () => {
        const tree = renderer.create(<ResetPassword onReset={jest.fn()} />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})