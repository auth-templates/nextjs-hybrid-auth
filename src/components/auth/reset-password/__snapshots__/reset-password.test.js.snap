// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword renders correctly 1`] = `
<div
  className="resetPassword"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="group"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <div
        className="passwordInput"
      >
        <input
          className="normal input"
          id="password"
          name="password"
          onFocus={[Function]}
          required={true}
          type="password"
        />
        <button
          className="visibilityIcon"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 6c3.79 0 7.17 2.13 8.82 5.5C19.17 14.87 15.79 17 12 17s-7.17-2.13-8.82-5.5C4.83 8.13 8.21 6 12 6m0-2C7 4 2.73 7.11 1 11.5 2.73 15.89 7 19 12 19s9.27-3.11 11-7.5C21.27 7.11 17 4 12 4zm0 5c1.38 0 2.5 1.12 2.5 2.5S13.38 14 12 14s-2.5-1.12-2.5-2.5S10.62 9 12 9m0-2c-2.48 0-4.5 2.02-4.5 4.5S9.52 16 12 16s4.5-2.02 4.5-4.5S14.48 7 12 7z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      className="group"
    >
      <label
        htmlFor="confirmPassword"
      >
        Confirm password
      </label>
      <div
        className="passwordInput"
      >
        <input
          className="normal input"
          id="confirmPassword"
          name="confirmPassword"
          onFocus={[Function]}
          required={true}
          type="password"
        />
        <button
          className="visibilityIcon"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 6c3.79 0 7.17 2.13 8.82 5.5C19.17 14.87 15.79 17 12 17s-7.17-2.13-8.82-5.5C4.83 8.13 8.21 6 12 6m0-2C7 4 2.73 7.11 1 11.5 2.73 15.89 7 19 12 19s9.27-3.11 11-7.5C21.27 7.11 17 4 12 4zm0 5c1.38 0 2.5 1.12 2.5 2.5S13.38 14 12 14s-2.5-1.12-2.5-2.5S10.62 9 12 9m0-2c-2.48 0-4.5 2.02-4.5 4.5S9.52 16 12 16s4.5-2.02 4.5-4.5S14.48 7 12 7z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      className="captcha"
    >
      <div
        className="captchaWraper"
      >
        <div
          className="captchaImage"
        >
          <div
            className="ldsRing"
            data-testid="small-lds-ring"
          >
            <div />
            <div />
          </div>
        </div>
        <button
          className="normal"
          onClick={[Function]}
        >
           Refresh 
        </button>
      </div>
      <input
        className="normal"
        id="captcha"
        name="catcha"
        placeholder="Enter captcha"
        required={true}
      />
    </div>
    <button
      className="normal resetButton"
      type="submit"
    >
      Reset password
    </button>
  </form>
</div>
`;
